File: frequency_colors.py
from util.database.frequency_color import fetch_frequency_color, update_frequency_color, add_frequency_color, remove_frequency_color
from util.database.database import Database


class FrequencyColors:
    def __init__(self, scene_id: int | None = None, color_sequence_id_1: int | None = None, 
                        color_sequence_id_2: int | None = None, color_sequence_id_3: int | None = None):
        self.scene_id = scene_id
        
        if scene_id != None and color_sequence_id_1 != None and color_sequence_id_2 != None and color_sequence_id_3 != None:
            self.color_sequence_id_1 = color_sequence_id_1
            self.color_sequence_id_2 = color_sequence_id_2
            self.color_sequence_id_3 = color_sequence_id_3
        
        elif self.scene_id != None:
            frequency_color_data = fetch_frequency_color(self.scene_id)

            if not frequency_color_data:
                self.sync_changes_to_db(new=True)
                
                self.color_sequence_id_1 = 0
                self.color_sequence_id_2 = 0
                self.color_sequence_id_3 = 0

            self.color_sequence_id_1 = frequency_color_data['color_sequence_id_1']
            self.color_sequence_id_2 = frequency_color_data['color_sequence_id_2']
            self.color_sequence_id_3 = frequency_color_data['color_sequence_id_3']
    
    def sync_changes_to_db(self, new: bool = False) -> bool:
        if new:
            return add_frequency_color(self.scene_id, self.color_sequence_id_1, 
                                       self.color_sequence_id_2, self.color_sequence_id_3)
        
        return update_frequency_color(self.scene_id, self.color_sequence_id_1, 
                                      self.color_sequence_id_2, self.color_sequence_id_3)

File: src/FrequencyColors.cpp
#pragma once

#include "FrequencyColors.hpp"

#include "util/database/FrequencyColorDB.hpp"
#include "util/database/Database.hpp"

FrequencyColors::FrequencyColors(std::optional<int> scene_id,
                                 std::optional<int> color_sequence_id_1,
                                 std::optional<int> color_sequence_id_2,
                                 std::optional<int> color_sequence_id_3)
    : scene_id(scene_id.value_or(-1))
{
    if (scene_id && color_sequence_id_1 && color_sequence_id_2 && color_sequence_id_3) {
        this->color_sequence_id_1 = color_sequence_id_1.value();
        this->color_sequence_id_2 = color_sequence_id_2.value();
        this->color_sequence_id_3 = color_sequence_id_3.value();
    }
    else if (scene_id) {
        auto freq_color_data = FrequencyColorDB::fetch_frequency_color(scene_id.value());

        this->color_sequence_id_1 = 0;
        this->color_sequence_id_2 = 0;
        this->color_sequence_id_3 = 0;
        this->sync_changes_to_db(true);
    }
}

bool FrequencyColors::sync_changes_to_db(bool is_new) {
    if (is_new) {
        return FrequencyColorDB::add_frequency_color(this->scene_id,
                                                     this->color_sequence_id_1,
                                                     this->color_sequence_id_2,
                                                     this->color_sequence_id_3);
    }

    return FrequencyColorDB::update_frequency_color(this->scene_id,
                                                    this->color_sequence_id_1,
                                                    this->color_sequence_id_2,
                                                    this->color_sequence_id_3);
}
